### ISSUE TITLE
The Newton diagnostics should not be run only on the first nonlinear system of the initialization problem

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13318]
Run this MWE involving the Newton Diagnostics
```modelica
model Test
  Real x(start = 0.5), y(start = 0.5);
  Real z(start = 0.5), w(start = 0.5);
equation
  z^3 + 1e-6*z + w^3 + 1e-6*w = 1;
  z^3 + 1e-6*z - w^3 + 1e-6*w = 0;
  x^3 + 1e-6*x + y^3 + 1e-6*y = 1;
  x^3 + 1e-6*x - y^3 + 1e-6*y = 0;
annotation(
    __OpenModelica_simulationFlags(lv = "LOG_NLS_NEWTON_DIAG", newtonDiagnostics = "()"),
    __OpenModelica_commandLineOptions = "--tearingMethod=minimalTearing");
end Test;
```
The following output is obtained:
```
   ========================================================
   Final results
   ========================================================

      alpha_i > 1e-02
      ------------------------
      alpha_2 =     0.333

      Gamma_ijk > 1e-02
      ------------------------
      Gamma_1_1_1 =     0.500
      Gamma_1_2_2 =     0.500
      Gamma_2_1_1 =     0.500
      Gamma_2_2_2 =     0.500

      |Sigma_jj| > 1e-02
      ------------------------
      Sigma_1_1 =     2.000
      Sigma_2_2 =     2.000

      Variables    Initial guess         |max(Gamma,Sigma)|
      ---------    ------------------    ------------------
      var_2           w =     0.5        2.000
      var_1           z =     0.5        2.000

      Equations    max(alpha,Gamma)
      ---------    ------------------
      eqn_1        0.500
      eqn_2        0.500

   ========================================================
```
The shown results obviously only refer to the first system of equation.

**Comment by casella** [comment]
My recommendation is to have two operating modes:
- `-newtonDiagnostics` runs the analysis on all the nonlinear systems of the initialization problem
- `-newtonDiagnostics=<index>` only runs it on the system specified by the index, as show in the Declarative Debugger

The second option allows to avoid unnecessarily long log files and to focus on the system that is causing the initialization failure.

**Comment by casella** [comment]
@phannebohm, if I now run this test case
```
loadFile("NewtonDiagnostics.mo");getErrorString();
simulate(NewtonDiagnostics.ThermoFluidSystem.M6);getErrorString();
```
on the attached package [NewtonDiagnostics.mo](https://github.com/user-attachments/files/18216833/NewtonDiagnostics.mo.txt)

I get this log lines:
```
LOG_NLS_NEWTON_DIAGNOSTICS | info    | Running newton diagnostics for system 13
LOG_NLS_NEWTON_DIAGNOSTICS | info    | Running newton diagnostics for system 26
```
which correspond to:
![immagine](https://github.com/user-attachments/assets/3984ca05-6af9-4b58-8eaf-60e3e0546036)

We should only run the diagnostics on the *initial equations* only, not on the regular ones. 

As @TeusvanderStelt noted, this is the root cause of  #13409: when we call the diagnostics on the regular equation, the variables are already initialized, so the delta's and the residuals are very small or zero, and this can crash the code when computing the indicators.

**Comment by casella** [comment]
This is kind of weird. I found this chunk of code:

https://github.com/OpenModelica/OpenModelica/blob/900ec0891f9ef6860dcf42547c60ba6a584eaf95/OMCompiler/SimulationRuntime/c/simulation/solver/nonlinearSystem.c#L1228-L1231

that apparently calls the `newtonDiagnostics()` function only if `data->simulationInfo->initial` is non-zero. I don't get why it also runs for regular equations.

@phannebohm what supposed to be contained in `data->simulationInfo->initial`?

**Comment by phannebohm** [comment]
Yes, exactly that is the part I changed in #13407 (you can see it more clearly if you hide whitespace changes).

> @phannebohm what supposed to be contained in `data->simulationInfo->initial`?

It's a boolean that is set to `TRUE` at the start of initialization and `FALSE at the end here (I'll fix that type issue in an upcoming PR):
https://github.com/OpenModelica/OpenModelica/blob/900ec0891f9ef6860dcf42547c60ba6a584eaf95/OMCompiler/SimulationRuntime/c/simulation/solver/initialization/initialization.c#L765


> Or maybe it is because regular equations are called once at the end of the initialization process, when that flag is still true?

This is what I suspect as well, not sure yet.

> We should instead base the decision on the attribute of the specific system of equation being analyzed.

I tried to find the attribute that the debugger is using to categorize regular/initial/... equations. Turns out it is not an attribute, instead during serialization the different equation lists are written into the *_info.json.
https://github.com/OpenModelica/OpenModelica/blob/900ec0891f9ef6860dcf42547c60ba6a584eaf95/OMCompiler/Compiler/SimCode/SerializeModelInfo.mo#L96-L109
So I guess we need to read that json file or store the info somewhere else.

**Comment by casella** [comment]
> > Or maybe it is because regular equations are called once at the end of the initialization process, when that flag is still true?
> 
> This is what I suspect as well, not sure yet.

That would be inconsistent with the comment in the source code
https://github.com/OpenModelica/OpenModelica/blob/900ec0891f9ef6860dcf42547c60ba6a584eaf95/OMCompiler/SimulationRuntime/c/simulation/solver/initialization/initialization.c#L765

But it would also be inconsistent with what the specification says.

As I understand, assuming there is no homotopy for simplicity, the runtime first solves the initialization problem, i.e. the regular equations with initial() returning true coupled with the initial equations and fixed = true attributes for variables. That is the initialization proper, and that's what we want to debug. At the end of this process, the initial values of states and parameters with fixed = false are known.

Then, the solution at time = StartTime is computed by solving the _regular_ equations starting from those state and parameter values, assuming initial() returns false.

This step can be considered as the last action of initialization or as the first action of simulation, it's really at the boundary of the two. 

This MWE demonstrates the whole story:
```modelica
model TestInitialBoundary
  Real x;
  Real y;
equation
  der(x) = y;
  y + (if initial() then 0 else 1e-12*sin(y)) = if initial() then x + 1e-12*sin(x) - 1 else x + sin(x) - 1;
initial equation
  der(x) = 0;
end TestInitialBoundary;
```
The initialization problem proper is:
```modelica
der(x) = y;
y = x + 1e-9*sin(x) - 1;
der(x) = 0;
```
whose solution is
```modelica
der(x) = 0;
y = 0;
x = 1; // with rtol < 1e-12
```
However, when the simulation starts at time = 0, the regular equations are solved with initial() returning false, hence
```modelica
y + 1e-12*sin(y) = x + sin(x) - 1;
der(x) = y;
```
Now, plugging in x = 1 and neglecting the 1e-12*sin(y) term, you get
```
x = 1;
y = 1 + sin(1) - 1 = 0.84
der(x) = y = 0.84
```
which is what you see in the simulation results at time = 0
![immagine](https://github.com/user-attachments/assets/4557a44b-5c4c-4137-b330-68f8cc0c1e36)

Now, this model has two nonlinear equations:

![immagine](https://github.com/user-attachments/assets/a1052303-2d91-40f7-b944-0b91bd9517dc)

only the first one should be analyzed. But the point is, as I understand from the output of the debugger in this case, conditional expressions with `if initial()` should be evaluated at compile time, not at run time.

**Comment by casella** [comment]
BTW, this simple script:
```
loadString("
model TestInitialBoundary
  Real x;
  Real y;
equation
  der(x) = y;
  y + (if initial() then 0 else 1e-12*sin(y)) = if initial() then x + 1e-12*sin(x) - 1 else x + sin(x) - 1;
initial equation
  der(x) = 0;
end TestInitialBoundary;");getErrorString();
simulate(TestInitialBoundary, simflags="-lv=LOG_NLS_NEWTON_DIAGNOSTICS -newtonDiagnostics");getErrorString();
```
currently gives this output

```
true
""
record SimulationResult
    resultFile = "",
    simulationOptions = "startTime = 0.0, stopTime = 1.0, numberOfIntervals = 500, tolerance = 1e-6, method = 'dassl', fileNamePrefix = 'TestInitialBoundary', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = '-lv=LOG_NLS_NEWTON_DIAGNOSTICS -newtonDiagnostics'",
    messages = "Simulation execution failed for model: TestInitialBoundary
LOG_NLS_NEWTON_DIAGNOSTICS | info    | Running newton diagnostics for system 4
LOG_NLS_NEWTON_DIAGNOSTICS | info    | Newton diagnostics starting ....
",
    timeFrontend = 0.0032213,
    timeBackend = 0.0110805,
    timeSimCode = 9.157e-4,
    timeTemplates = 0.010042,
    timeCompile = 1.9791329,
    timeSimulation = 4.5735481,
    timeTotal = 6.5781475
end SimulationResult;
""
```
i.e., it crashes for some reason while analyzing the first equation (?)

**Comment by casella** [comment]
> I tried to find the attribute that the debugger is using to categorize regular/initial/... equations. Turns out it is not an attribute, instead during serialization the different equation lists are written into the *_info.json.
> ...
> So I guess we need to read that json file or store the info somewhere else.

Yeah, I'd say reading the json file is the easiest option. One could:
- read in the json file
- get the "equations" name-value pair
- scan the array of the value part to build a list of initial non-linear equations and initial-lambda0 non-linear equations, i.e. identify those array members that contain the "display":"non-linear" pair and either the "section":"initial" or the "section":"initial-lambda0" pair and add the index paired to the "eqIndex" key to the list. The list can be stored in a simple array of int indeces
- set the list of candidate equations for diagnostics to the initial-lambda0 list if non-empty, otherwise to the initial list
- if no -lv_system flag is set, call newtonDiagnostics() on each system in the list of candidates, otherwise call newtonDiagnostics() on the intersection of the list of candidates and the list of indeces in the -lv_system flag

This would also resolve #13318
I have no idea how to process json files, but I guess with the right tools (that we already use in the runtime, BTW) what I wrote above would be doable with a couple dozen lines of code. BTW, the runtime already uses information from the json file, e.g. to get the parameter values, so I guess the required data structures are already available in the runtime code.

 @phannebohm, are you familiar enough with json stuff to take care of that, or to give instructions to @TeusvanderStelt to do it?

**Comment by casella** [comment]
@phannebohm needs to look this up himself anyway, so he'll take care of that himself. 

It seems that [simulation_info_json.c](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/simulation/simulation_info_json.c) is the right place to look for, in particular

https://github.com/OpenModelica/OpenModelica/blob/14fee190889d353e697f003218a48f844d87339f/OMCompiler/SimulationRuntime/c/simulation/simulation_info_json.c#L381-L388

reads the data from the equations array into a string, while

https://github.com/OpenModelica/OpenModelica/blob/14fee190889d353e697f003218a48f844d87339f/OMCompiler/SimulationRuntime/c/simulation/simulation_info_json.c#L299-L308

reads the data about a specific equation with index `i`. Just call it in a for loop, et voilà 😃

**Comment by casella** [comment]
I checked with the new nightly build and the [NewtonDiagnostics.mo](https://github.com/user-attachments/files/18466560/NewtonDiagnostics.mo.txt) test package:
- only initial equations are checked (NewtonDiagnostics.ThermoFluidSystem.M5)
- only initial-lambda0 equations are checked if there are equations using homotopy (NewtonDiagnostics.Tests.TestHomotopy)
- only the systems selected with -lv_systems are checked (NewtonDiagnostics.Tests.TestSelectSystem)

This ticket can be closed. Thanks @phannebohm!
